File: scripts/github_sim.py
import hmac
import hashlib
import urllib.request
import json
import os

# 1. Читаем секрет из файла .env напрямую
env_secret = "test-local-secret-key-123" # Вставь сюда значение WEBHOOK_SECRET из твоего .env файла
if os.path.exists(".env"):
    with open(".env", "r") as f:
        for line in f:
            if line.startswith("WEBHOOK_SECRET="):
                env_secret = line.strip().split("=", 1)[1]

# 2. Фейковый payload (Вася делает коммит и случайно пушит файл с секретами)
payload = json.dumps({
    "action": "push",
    "repository": {"name": "frontend-app"},
    "sender": {"login": "vasya_developer"},
    "commits": [
        {
            "id": "a1b2c3d4",
            "message": "Update config and fix bugs",
            "added": ["config/database.yml"],   # <-- Этот файл мы будем "сканировать"
            "modified": ["src/main.js"]
        }
    ]
}).encode('utf-8')

# 3. Вычисляем эталонную подпись так же, как это делает настоящий GitHub
signature = hmac.new(
    env_secret.encode('utf-8'),
    payload,
    hashlib.sha256
).hexdigest()

# 4. Формируем заголовки запроса
headers = {
    'Content-Type': 'application/json',
    'X-Hub-Signature-256': f'sha256={signature}',
    'X-GitHub-Delivery': 'dummy-delivery-id-777'
}

# 5. Отправляем POST-запрос на наш Go-шлюз
req = urllib.request.Request('http://localhost:8080/api/v1/webhook', data=payload, headers=headers)

try:
    with urllib.request.urlopen(req) as response:
        print(f"✅ УСПЕХ! Статус: {response.getcode()}")
        print(f"Ответ сервера: {response.read().decode('utf-8')}")
except urllib.error.HTTPError as e:
    print(f"❌ ОШИБКА! Статус: {e.code}")
    print(f"Ответ сервера: {e.read().decode('utf-8')}")

File: scripts/github_sim.go
//go:build ignore

package main

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"net/http"
	"time"
)

const (
	targetURL = "http://localhost:8080/webhook"
	secret    = "whsec_fbf69a17e74e7a6eb9a169f12904a5edcca6aa77d9d39638afc2f6fcb3c39d1a" // Синхронизировано с fallback-значением в main.go
)

func main() {
	// 1. Формирование синтетического Payload (Selective Unmarshaling test)
	payload := []byte(`{
		"action": "opened",
		"repository": {
			"full_name": "XtReL/devsecops-gatekeeper"
		},
		"sender": {
			"login": "XtReL"
		},
		"pull_request": {
			"number": 42
		},
		"malicious_injection": "this_should_be_ignored"
	}`)

	// 2. Вычисление математически корректной подписи (Bypass Ring 0)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	signature := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	// 3. Транспорт (HTTP POST)
	req, err := http.NewRequest("POST", targetURL, bytes.NewBuffer(payload))
	if err != nil {
		fmt.Printf("sre_alert: request compilation failed: %v\n", err)
		return
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Hub-Signature-256", signature)

	fmt.Printf(">>> Sending payload to %s\n", targetURL)
	fmt.Printf(">>> Signature: %s\n", signature)

	// 4. Исполнение
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("sre_alert: connection refused. Is Gatekeeper running? Error: %v\n", err)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("<<< Gatekeeper Status: %s\n", resp.Status)
	fmt.Printf("<<< Gatekeeper Body: %s\n", string(body))
}
